Add documentation for customfields post and put

evanhsu opened the pull request:
- Add documentation for the `POST /customfields` operation (create new customfields).
- Add documentation for the `PUT /customfields` operation (update existing customfields).

Commit:
Added docs for customfield create (POST )

Commit:
Added docs for the PUT /customfields operation (Update customfields)

dustinsweet commented on source/includes/APIReference/CustomFields/Examples/_create-request.md.erb:
```diff
@@ -0,0 +1,13 @@
+{
```
Indent 2 spaces (as in the update example) for better use of limited real estate.

evanhsu replied:
Ok, I think the indentation matches now 👍

dustinsweet reviewed (changes requested):
Excited to see this!  Thanks Evan

Commit:
Update the `_create-response.md.erb` example to show that results are numerically indexed instead of indexed by id

Commit:
Update indentation in the customfields 'create-request' example to be consistent with the other examples.

Commit:
Update indentation in the customfields 'create-response' example to be consistent with the other examples.

tylerrhinehart commented on source/includes/APIReference/CustomFields/_create.md.erb:
```diff
@@ -0,0 +1,40 @@
+| **show_to_all**<br/>optional | _Boolean_ | Default: `true`.  Declares whether this customfield should exist on timesheets for all workers. If false, it will only appear on timesheets for workers to whom it has been assigned. |
+| **short_code**<br/>optional | _String_ | This is a shortened code or alias that is associated with the customfield. It may only consist of letters and numbers.
+
+For a full list of the properties that may be set on a customfield, see [The Custom Field Object](#the-custom-field-object).
```
I think this may be a little confusing because all of the properties listed above are all the properties that aren't read-only. When I read this it made me think that there are more properties I can pass in.

evanhsu replied:
I'll test out some different verbiage here, maybe:

> For a full list of properties that exist on a customfield, see [The Custom Field Object](#the-custom-field-object).

jaredc replied:
Yeah, I like your suggested change, @evanhsu

EldinTurulja replied:
Would we also want to add `show_to_all` to the Custom Field Object, since from my understanding, and how I coded it (subject to change), it should be editable on the PUT as well?

evanhsu replied:
I've added the `show_to_all` property to the Custom Field Object as a `read-write` property

tylerrhinehart commented on source/includes/APIReference/CustomFields/_create.md.erb:
```diff
@@ -0,0 +1,40 @@
+| **active**<br/>optional | _Boolean_ | If false, this `customfield` is considered archived and will not be visible or usable. |
+| **applies_to**<br/>required | _String_ | Allowed values: 'timesheet'. Specify that this `customfield` should appear on timesheets (additional customfield types may be supported in the future). |
+| **type**<br/>required | _String_ | Allowed values: 'managed-list', 'free-form'.  A managed-list customfield is intended to be displayed as a dropdown list, where each 'option' is a customfielditem. A 'free-form' customfield supports text entry and does not require a collection of customfielditems to accompany it. |
+| **show_to_all**<br/>optional | _Boolean_ | Default: `true`.  Declares whether this customfield should exist on timesheets for all workers. If false, it will only appear on timesheets for workers to whom it has been assigned. |
```
For the first pass at this, `show_to_all` is going to be `global` instead to match the supplemental_data
```suggestion
| **global**<br/>optional | _Boolean_ | Default: `true`.  Declares whether this customfield should exist on timesheets for all workers. If false, it will only appear on timesheets for workers to whom it has been assigned. |
```

tylerrhinehart commented on source/includes/APIReference/CustomFields/Examples/_update-response.md.erb:
```diff
@@ -0,0 +1,40 @@
+        "type": "managed-list",
+        "short_code": "t",
+        "regex_filter": "",
+        "show_to_all": true,
```
```suggestion
        "global": true,
```

tylerrhinehart commented on source/includes/APIReference/CustomFields/Examples/_update-response.md.erb:
```diff
@@ -0,0 +1,40 @@
+        "type": "managed-list",
+        "short_code": "vin",
+        "regex_filter": "",
+        "show_to_all": true,
```
```suggestion
        "global": true,
```

tylerrhinehart commented on source/includes/APIReference/CustomFields/Examples/_create-response.md.erb:
```diff
@@ -0,0 +1,23 @@
+        "type": "managed-list",
+        "short_code": "vin",
+        "regex_filter": "",
+        "show_to_all": true,
```
```suggestion
        "global": true,
```

tylerrhinehart commented on source/includes/APIReference/CustomFields/Examples/_create-request.md.erb:
```diff
@@ -0,0 +1,14 @@
+      "active": true,
+      "applies_to": "timesheet",
+      "type": "managed-list",
+      "show_to_all": true,
```
```suggestion
      "global": true,
```

tylerrhinehart reviewed (commented):
We have decided to initially release this with the `global` property instead of `show_to_all` to keep it consistent and backward-compatible with the supplemental_data.

evanhsu committed:
Rename 'show_to_all' to 'global'

Co-Authored-By: Tyler Rhinehart <trhinehart@wps-inc.com>

evanhsu reviewed (commented):
One thing to note here for discussion:

We're going to expose the `global` property in the response to a POST or PUT request (because we're allowing that property to be set), but it doesn't appear in the response to a GET request.  Should we add it to the GET response now that we support PUT and POST?

tylerrhinehart commented:
Disregard all my previous comments about the global property. That value should now be exposed as `show_to_all`.

Commit:
Revert "Rename 'show_to_all' to 'global'"

This reverts commit 9c17bc505cd47e1f0b9df61ae635dc372cb9b9a9.
The net effect is that the 'global' property is now called 'show_to_all'
(again).

evanhsu commented:
I've reverted the name of the `global` fields back to `show_to_all`

jaredc commented on source/includes/APIReference/CustomFields/_create.md.erb:
```diff
@@ -0,0 +1,40 @@
+|                |             |             |
+| -------------: | :---------: | ----------- |
+| **name**<br/>required | _String_ | The name of this customfield. |
+| **required**<br/>optional | _Boolean_ | If true, a non-null value must be selected for this customfield when it appears on a timesheet. |
```
Could you include what the default value is for any of these optional parameters are?  So that the consumer will know what'll happen if they omit it.

EldinTurulja replied:
Seconding this idea, I can get this data for you if needed.

evanhsu replied:
I've added default values now

jaredc commented on source/includes/APIReference/CustomFields/_create.md.erb:
```diff
@@ -0,0 +1,40 @@
+
+|         |          |
+| :-----: | :------- |
+| <code class="level200">200</code> | OK. customfield was created successfully. |
```
In the case of creating a customfield with a name that already is tied to an archived customfield, we actually do not create one, but instead unarchive the archived one.  I'm thinking we should give a status back of 202 in that case, similar to what we do for Geolocations.  

```
| <code class="level200">202</code> | OK. customfield was accepted, but not created (because we already have a customfield with that name. We've unarchived it for you.) |
```

EldinTurulja replied:
Did some manual testing and found that through the UI, it actual stops you from creating the name at all if it is in use, even if it is used by an archived custom field. The current implementation in the API was to rename the archived one with a date stamp appended at the end, and create the new one with the desired name. Is this difference acceptable or do we try to mirror functionality?

evanhsu replied:
Ok, I'm going to leave the docs as they are with the `200` response since the API will not provide the same side-effects that the web UI provides.

jaredc commented on source/includes/APIReference/CustomFields/_create.md.erb:
```diff
@@ -0,0 +1,40 @@
+| **active**<br/>optional | _Boolean_ | If false, this `customfield` is considered archived and will not be visible or usable. |
+| **applies_to**<br/>required | _String_ | Allowed values: 'timesheet'. Specify that this `customfield` should appear on timesheets (additional customfield types may be supported in the future). |
+| **type**<br/>required | _String_ | Allowed values: 'managed-list', 'free-form'.  A managed-list customfield is intended to be displayed as a dropdown list, where each 'option' is a customfielditem. A 'free-form' customfield supports text entry and does not require a collection of customfielditems to accompany it. |
+| **show_to_all**<br/>optional | _Boolean_ | Default: `true`.  Declares whether this customfield should exist on timesheets for all workers. If false, it will only appear on timesheets for workers to whom it has been assigned. |
```
We don't support filtering/assignment based on workers for _fields_, only for items.  For fields, this is only controlled based on jobcode.  This 'show_to_all' setting corresponds to the 'Show for all customers' setting in the UI.  Hmm... perhaps we should rename this to 'show_for_all_jobcodes' ?

So apart from the name of the property, I suggest we change this to:

```suggestion
| **show_to_all**<br/>optional | _Boolean_ | Default: `true`.  Declares whether this customfield should be shown on timesheets regardless of the jobcode chosen. If false, it will only appear when the chosen jobcode for a timesheet has been associated with this field. |
```

jsalazar7 replied:
Second this comment, the filtering applies to job codes, not workers.

jaredc reviewed (commented):
Looks good, Evan.  Thank you for doing this!

See my comments for a few things that need to change.  I'll leave final approval to someone else.  Let me know if you want to discuss any of my comments.

EldinTurulja reviewed (commented):
Nice work! Just a question about `show_to_all`, I believe it should be in the CF Object.

dustinsweet commented:
Please update the changelog file

jsalazar7 reviewed (changes requested):
This looks good Evan. I just think that the phrase "If false, it will only appear on timesheets for workers to whom it has been assigned." should be made specific to job codes and not workers.

evanhsu committed:
Update source/includes/APIReference/CustomFields/_create.md.erb

Fix description of the 'show_to_all' field to reference the jobcode rather than the user

Co-Authored-By: Jared Cheney <jared.cheney@gmail.com>

Commit:
Updated the changelog

Commit:
Added default values for all optional properties in the 'create' docs.

Commit:
Add the `show_to_all` property to the `customfield` object docs because it's now a 'read-write' property.
Change the `type` property to be read-only.

Commit:
Updated the main header for the custom fields object to mention that a max of 6 can be active.

Commit:
Update the type property to be read-only (instead of just 'read')

jsalazar7 commented:
The updates look good!

EldinTurulja commented on source/includes/APIReference/CustomFields/_create.md.erb:
```diff
@@ -0,0 +1,39 @@
+| **active**<br/>optional | _Boolean_ | Default: `true`. If false, this `customfield` is considered archived and will not be visible or usable. |
+| **applies_to**<br/>required | _String_ | Default: `timesheet`. Allowed values: 'timesheet'. Specify that this `customfield` should appear on timesheets (additional customfield types may be supported in the future). |
+| **type**<br/>required | _String_ | Allowed values: 'managed-list', 'free-form'.  A managed-list customfield is intended to be displayed as a dropdown list, where each 'option' is a customfielditem. A 'free-form' customfield supports text entry and does not require a collection of customfielditems to accompany it. |
+| **show_to_all**<br/>optional | _Boolean_ | Default: `true`.  Declares whether this customfield should be shown on timesheets regardless of the jobcode chosen. If false, it will only appear when the chosen jobcode for a timesheet has been associated with this field. |
```
This defaults to false.

evanhsu replied:
```suggestion
| **show_to_all**<br/>optional | _Boolean_ | Default: `false`.  Declares whether this customfield should be shown on timesheets regardless of the jobcode chosen. If false, it will only appear when the chosen jobcode for a timesheet has been associated with this field. |
```

EldinTurulja commented on source/includes/APIReference/CustomFields/_create.md.erb:
```diff
@@ -0,0 +1,39 @@
+|                |             |             |
+| -------------: | :---------: | ----------- |
+| **name**<br/>required | _String_ | The name of this customfield. |
+| **required**<br/>optional | _Boolean_ | Default: `true`. If true, a non-null value must be selected for this customfield when it appears on a timesheet. |
```
This defaults to false.

evanhsu replied:
Hey dude, this is wrong. Try this instead:
```suggestion
| **required**<br/>optional | _Boolean_ | Default: `false`. If true, a non-null value must be selected for this customfield when it appears on a timesheet. |

```

EldinTurulja commented on source/includes/APIReference/CustomFields/_create.md.erb:
```diff
@@ -0,0 +1,39 @@
+| **applies_to**<br/>required | _String_ | Default: `timesheet`. Allowed values: 'timesheet'. Specify that this `customfield` should appear on timesheets (additional customfield types may be supported in the future). |
+| **type**<br/>required | _String_ | Allowed values: 'managed-list', 'free-form'.  A managed-list customfield is intended to be displayed as a dropdown list, where each 'option' is a customfielditem. A 'free-form' customfield supports text entry and does not require a collection of customfielditems to accompany it. |
+| **show_to_all**<br/>optional | _Boolean_ | Default: `true`.  Declares whether this customfield should be shown on timesheets regardless of the jobcode chosen. If false, it will only appear when the chosen jobcode for a timesheet has been associated with this field. |
+| **short_code**<br/>optional | _String_ | Default: `""` (none). This is a shortened code or alias that is associated with the customfield. It may only consist of letters and numbers.
```
Could be good to point out if nothing is specified here, we auto generate one for them.

EldinTurulja commented on source/includes/APIReference/CustomFields/_create.md.erb:
```diff
@@ -0,0 +1,39 @@
+
+<%= partial "includes/APIReference/CustomFields/Examples/create.tmpl.erb" %>
+
+Add one or more customfields.
```
Could be good to note here, or in the 'active' property: you can only have 6 active fields at a time.

evanhsu replied:
We discussed this in-person. Decided to leave this information in the header block for the customfield object itself.

Commit:
Update short_code description to mention that one will be autogenerated

evanhsu committed:
Correct some defaults

Commit:
Merge branch 'feat/add-customfield-post-and-put' of github.com:TSheetsTeam/api_docs into feat/add-customfield-post-and-put

# Conflicts:
#	source/includes/APIReference/CustomFields/_create.md.erb

EldinTurulja reviewed (approved):
Did an over the shoulder for a couple changes to the default values. Looks good to ship.
